tensorflow/python/eager/context.py

```diff
@@ -1244,6 +1244,31 @@ def invoking_op_callbacks(self):
   def invoking_op_callbacks(self, value):
     self._thread_local_data.invoking_op_callbacks = value
 
+  # Parse PhysicalDevice from the device string.
+  # If the device string contains subdevice type string, then parse it.
+  #  e.g. '/physical_device:GPU:0:X_GPU' 
```

Comment:
It seems better to put device index at the end.
`'/physical_device:GPU:0:X_GPU' `  --> `'/physical_device:GPU:X_GPU:0'`
The order of components is hierarchical in some way. physical_device > GPU > X_GPU > X_GPU device index.

Reply:
This order is following current [cpu](https://github.com/tensorflow/tensorflow/blob/89c2f7f062ee125bd3783035b1be07b5c0f9ea73/tensorflow/core/common_runtime/threadpool_device_factory.cc#L32)/[gpu](https://github.com/tensorflow/tensorflow/blob/89c2f7f062ee125bd3783035b1be07b5c0f9ea73/tensorflow/core/common_runtime/gpu/gpu_device.cc#L1063)/..etc convention, when splited with ":", the third part is the id. And I think this is more align with user behavior to specify the device(device_type:device_id)

Reply:
I generally think that TYPE:SUB_TYPE precisely define the type. So, `'/physical_device:GPU:X_GPU:0'` in this case is of the form device_type:device_id although.

Adding a few people to see if they have an opinion on this.
+ @sanjoy @rohan100jain

Reply:
I too prefer `/physical_device:GPU:X_GPU:0`.  If there are places that hardcode "when splited with ":", the third part is the id" then we should fix those to use `DeviceNameUtils` instead, and change `DeviceNameUtils` to understand the `/physical_device:GPU:X_GPU:0` format.

Reply:
@annarev @sanjoy  thanks for the comments, I have changed as your suggestion. DeviceNameUtils  will parse the front-end string, not this string(provided by DeviceFactroy, so it has no this issue.)